src/Symfony/Component/Serializer/Encoder/CsvEncoder.php

```diff
@@ -30,20 +30,45 @@ class CsvEncoder implements EncoderInterface, DecoderInterface
     const ESCAPE_FORMULAS_KEY = 'csv_escape_formulas';
     const AS_COLLECTION_KEY = 'as_collection';
 
-    private $delimiter;
-    private $enclosure;
-    private $escapeChar;
-    private $keySeparator;
-    private $escapeFormulas;
     private $formulasStartCharacters = array('=', '-', '+', '@');
+    private $defaultContext = array(
+        self::DELIMITER_KEY => ',',
+        self::ENCLOSURE_KEY => '"',
+        self::ESCAPE_CHAR_KEY => '\\',
+        self::KEY_SEPARATOR_KEY => '.',
+        self::ESCAPE_FORMULAS_KEY => false,
+    );
 
-    public function __construct(string $delimiter = ',', string $enclosure = '"', string $escapeChar = '\\', string $keySeparator = '.', bool $escapeFormulas = false)
+    /**
+     * @param array  $defaultContext
+     * @param string $enclosure
+     * @param string $escapeChar
+     * @param string $keySeparator
+     * @param bool   $escapeFormulas
+     */
+    public function __construct($defaultContext = array(), string $enclosure = '"', string $escapeChar = '\\', string $keySeparator = '.', bool $escapeFormulas = false)
     {
-        $this->delimiter = $delimiter;
-        $this->enclosure = $enclosure;
-        $this->escapeChar = $escapeChar;
-        $this->keySeparator = $keySeparator;
-        $this->escapeFormulas = $escapeFormulas;
+        if (!\is_array($defaultContext)) {
+            @trigger_error(sprintf('The "delimiter" parameter is deprecated since Symfony 4.2, use the "%s" key of the context instead.', static::DELIMITER_KEY), E_USER_DEPRECATED);
+
+            $defaultContext = array(static::DELIMITER_KEY => $defaultContext);
+        }
+
+        $args = array(
+            array('enclosure', static::ENCLOSURE_KEY, '"'),
+            array('escapeChar', static::ESCAPE_CHAR_KEY,  '\\'),
+            array('keySeparator', static::KEY_SEPARATOR_KEY, '.'),
+            array('escapeFormulas', static::ESCAPE_FORMULAS_KEY, false),
+        );
+        foreach ($args as $arg) {
+            $val = ${$arg[0]};
+            if (${$arg[0]} !== $arg[2]) {
+                $this->defaultContext[$arg[1]] = $val;
+                @trigger_error(sprintf('The "%s" parameter is deprecated since Symfony 4.2, use the "%s" key of the context instead.', $arg[0], $arg[1]), E_USER_DEPRECATED);
```

Comment:
instead of this complex logic, I'd suggest to now deal with mixed signatures: either the 1st arg is an array - or not and then we trigger one deprecation and we build the defaultcontext out of the arguments  (like done in JsonDecode)